pysrc/bytewax/execution.py

```diff
@@ -39,68 +146,208 @@ def run(flow: Dataflow, inp: Iterable[Tuple[int, Any]]) -> List[Tuple[int, Any]]
-        return out.append
+    coro = run_main(flow, input_builder, output_builder)
 
-    _run(flow, input_builder, output_builder)
+    work_remains = True
+    while work_remains:
+        work_remains = _step_coro(coro)
 
-    return out
+        for epoch, item in out_buffer:
+            yield epoch, item
+        out_buffer.clear()
+
+
+def cluster_main(
+    flow: Dataflow,
+    input_builder: Callable[[int, int], Iterable[Tuple[int, Any]]],
+    output_builder: Callable[[int, int], Callable[[Tuple[int, Any]], None]],
+    addresses: Iterable[str],
+    proc_id: int,
+    worker_count_per_proc: int = 1,
+    proc_should_stop: Callable[[], bool] = lambda: False,
+) -> None:
+    """Execute a dataflow in the current process as part of a cluster.
+
+    You have to coordinate starting up all the processes in the
+    cluster and ensuring they each are assigned a unique ID and know
+    the addresses of other processes. You'd commonly use this for
+    starting processes as part of a Kubernetes cluster.
+
+    Blocks until execution is complete.
+
+    >>> flow = Dataflow()
+    >>> flow.capture()
+    >>> def input_builder(worker_index, worker_count):
+    ...     return enumerate(range(3))
+    >>> async def output_builder(worker_index, worker_count, epoch_items):
+    ...     async for epoch, item in epoch_items:
+    ...         print(epoch, item)
+    >>> cluster_main(
+    ...     flow,
+    ...     input_builder,
+    ...     output_builder,
+    ...     addresses=["localhost:2101"],
+    ...     proc_id=0,
+    ... )  # doctest: +ELLIPSIS
+    (...)
+
+    See `bytewax.run_cluster()` for a convenience method to pass data
+    through a dataflow for notebook development.
+
+    See `bytewax.spawn_cluster()` for starting a simple cluster
+    locally on one machine.
+
+    Args:
+
+        flow: Dataflow to run.
+
+        input_builder: Returns input that each worker thread should
+            process.
+
+        output_builder: Returns a callback function for each worker
+            thread, called with `(epoch, item)` whenever and item
+            passes by a capture operator on this process.
+
+        addresses: List of host/port addresses for all processes in
+            this cluster (including this one).
+
+        proc_id: Index of this process in cluster; starts from 0.
+
+        worker_count_per_proc: Number of worker threads to start on
+            this process.
+
+        proc_should_stop: Returns if workers in this process should
+            gracefully shutdown soon.
+
+    """
+    builders, comms_threads = WorkerBuilder.cluster(
+        addresses, proc_id, worker_count_per_proc
+    )
+
+    worker_raised_ex = threading.Event()
+
+    def worker_should_stop():
+        return proc_should_stop() or worker_raised_ex.is_set()
+
+    def worker_main(builder):
+        coro = builder.build(flow, input_builder, output_builder, worker_should_stop)
+        asyncio.run(WorkerCoroWrapper(coro))
```

Comment:
We were having trouble with a Rust `SendWrapper` type being accessed from a different thread. I __think__ this is the Python GC running on the main thread cleaning up the `WorkerCoro` but I'm not sure.